src/singleton.cr

```diff
@@ -0,0 +1,40 @@
+# The Singleton module implements the Singleton pattern.
+#
+### Usage
+#
+# ```
+# require "singleton"
+#
+# class Foo
+#   include Singleton
+# end
+#
+# foo = Foo.instance
+# ```
+
+module Singleton
+  macro included
+    @@mutex = Mutex.new
+
+    # Instantiate the object if it doesn't exist.  Always returns the same object.
+    def self.instance
+      @@instance ||= @@mutex.synchronize do
+        @@instance = new
+      end
```

Comment:
The comparison needs to be repeated inside the lock, else you can have the following situation:

```
A: @@instance || @@instance = @@mutex.synchronize do # @@instance = nil, enter lock, interrupt
B: @@instance || @@instance = @@mutex.synchronize do # @@instance = nil, enter lock, blocked
A: @@instance = new # set, var to instance, leave lock
A: do something with instance # interrupt
B: @@instance = new # lock released, override with new instance since missing check.
```